python/lsst/daf/butler/queries/_general_query_results.py

```diff
@@ -0,0 +1,125 @@
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("GeneralQueryResults",)
+
+from collections.abc import Iterator
+from typing import Any, final
+
+from .._dataset_ref import DatasetRef
+from .._dataset_type import DatasetType
+from ..dimensions import DataCoordinate, DimensionGroup
+from ._base import QueryResultsBase
+from .driver import QueryDriver
+from .result_specs import GeneralResultSpec
+from .tree import QueryTree, ResultColumn
+
+
+@final
+class GeneralQueryResults(QueryResultsBase):
+    """A query for `DatasetRef` results with a single dataset type.
+
+    Parameters
+    ----------
+    driver : `QueryDriver`
+        Implementation object that knows how to actually execute queries.
+    tree : `QueryTree`
+        Description of the query as a tree of joins and column expressions. The
+        instance returned directly by the `Butler._query` entry point should be
+        constructed via `make_unit_query_tree`.
+    spec : `GeneralResultSpec`
+        Specification of the query result rows, including output columns,
+        ordering, and slicing.
+
+    Notes
+    -----
+    This class should never be constructed directly by users; use `Query`
+    methods instead.
+    """
+
+    def __init__(self, driver: QueryDriver, tree: QueryTree, spec: GeneralResultSpec):
+        spec.validate_tree(tree)
+        super().__init__(driver, tree)
+        self._spec = spec
+
+    def __iter__(self) -> Iterator[dict[ResultColumn, Any]]:
+        """Iterate over result rows.
+
+        Yields
+        ------
+        row_dict : `dict` [`ResultColumn`, `Any`]
+            Result row as dictionary, the keys are `ResultColumn` instances.
+        """
+        for page in self._driver.execute(self._spec, self._tree):
+            columns = tuple(page.spec.get_result_columns())
+            for row in page.rows:
+                yield dict(zip(columns, row))
+
+    def iter_refs(self, dataset_type: DatasetType) -> Iterator[tuple[DatasetRef, dict[ResultColumn, Any]]]:
+        """Iterate over result rows and return DatasetRef constructed from each
+        row and an original row.
```

Comment:
There's one place in QG generation where we'd want a method sort of like this that returns a tuple of `DatasetRef` and `DataCoordinate` (it currently calls `DataCoordinateQueryResults.findRelatedDatasets`).

I had originally planned to do that via a special dict-like row object that had single-row `get_ref`, `get_data_id` etc. methods, but I think I like your iteration adapter better as it allows for doing work common to all rows up front.  But I'd like to extend this to be able to support at least `DatasetRef` + `DataCoordinate` + [extra stuff] somehow, and maybe even support `DatassetRef` objects for multiple dataset.  Maybe something like this?
```
def iter_tuples(self, *dataset_type: DatasetType) -> Iterator[tuple[DataCoordinate, dict[str, DatasetRef], dict[ResultColumn, Any]]
```
where it always returns the full data ID, a dict of `DatasetRef` keyed by dataset type name, and then whatever columns aren't included in one of those?  To be honest I'm not thrilled with this, either, but maybe it'll inspire you to find something better?

Reply:
The return annotation is hard to read. Would it make sense to introduce a helper NamedTuple like this:
```
class GeneralResultTuple(typing.NamedTuple):

    data_id: DataCoordinate
    refs: dict[str, DtasetRef]  # or maybe just list[DatasetRef]?
    raw_row: dict[str, Any]

class GeneralQueryResults(QueryResultsBase):

    def iter_tuples(self, *dataset_type: DatasetType) -> Iterator[GeneralResultTuple]:
        ...
```

Reply:
:+1: I do like `NamedTuple` for this; you can still unpack it if you want to, but it's also very clear what's in it and you can hang documentation on it.  And I like `list` rather than `dict` for the `DatasetRefs`, too, since we can make that ordering consistent with the argument rather than arbitrary.